app/services/role_checker.rb

```diff
@@ -0,0 +1,48 @@
+class RoleChecker
+  attr_reader :user, :records
+
+  EDITOR_ROLES = %w(owner collaborator expert scientist moderator)
+  APPROVER_ROLES = %w(owner collaborator)
+  VIEWER_ROLES = %w(owner collaborator expert scientist tester)
+
+  def initialize(user, records)
+    @user = user
+    @records = records
+    @viewer_project_ids = viewer_project_ids
```

Comment:
this memoized value will never be served as you don't have an `attr_reader` set for it (also that would clash with the method name)